android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/BridgeManagerProvider.java

```diff
@@ -1,52 +1,151 @@
 package org.sagebionetworks.bridge.android.manager;
 
+import android.app.Application;
 import android.content.Context;
 import android.support.annotation.AnyThread;
 import android.support.annotation.NonNull;
 
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+
 import org.sagebionetworks.bridge.android.BridgeConfig;
-import org.sagebionetworks.bridge.android.manager.auth.AuthenticationManager;
+import org.sagebionetworks.bridge.rest.ApiClientProvider;
 
 import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
 
 /**
- * Created by jyliu on 1/20/2017.
+ * This class contains configuration and performs setup of dependencies for a Bridge Application.
+ * <p>
+ * A call to {@link #getInstance()} returns a BridgeManagerProvider instance that is usable
+ * globally. The managers made available by BridgeManagerProvider maintain and share state across
+ * the Application.
  */
 @AnyThread
 public class BridgeManagerProvider {
     private static BridgeManagerProvider instance;
 
+    /**
+     * @return singleton instance
+     */
     @NonNull
-    public static BridgeManagerProvider getInstance(@NonNull Context context) {
-        if (instance == null) {
-            instance = new BridgeManagerProvider(new BridgeConfig(context));
-        }
+    public static BridgeManagerProvider getInstance() {
+        checkState(instance != null, "BridgeManagerProvider has not been initialized. " +
+                "Call init(Context) in your Application#onCreate()");
 
         return instance;
     }
 
+    /**
+     * Allows injection of singleton instance for testing purposes.
+     *
+     * @param bridgeManagerProvider instance to be returned by {@link #getInstance()}
+     */
+    public static void init(@NonNull BridgeManagerProvider bridgeManagerProvider) {
+        checkNotNull(bridgeManagerProvider);
+        instance = bridgeManagerProvider;
+    }
+
+    /**
+     * Intended to be called in {@link Application#onCreate()}
+     *
+     * @param applicationContext application's global context
+     */
+    public static void init(@NonNull Context applicationContext) {
+        checkNotNull(applicationContext);
+        checkState(instance == null, "BridgeManagerProvider has already been initialized");
+
+        instance = new BridgeManagerProvider(applicationContext);
+    }
+
+    private BridgeManagerProvider(@NonNull Context applicationContext) {
+        this.applicationContext = applicationContext;
+        bridgeConfig = new BridgeConfig(this.applicationContext);
```

Comment:
BridgeManagerProvider is a singleton attached to the application context, so right now I'm initializing BridgeManagerProvider in Application's onCreate(). I'm memoizing because onCreate() specifically suggests lazy loading: https://developer.android.com/reference/android/app/Application.html#onCreate()